File: clients/ruby/model/item.rb
require_relative 'weapon_type'
class Item
    def self.read_from(stream)
        discriminant = stream.read_int()
        if discriminant == Item::HealthPack::TAG
            return Item::HealthPack.read_from(stream)
        end
        if discriminant == Item::Weapon::TAG
            return Item::Weapon.read_from(stream)
        end
        if discriminant == Item::Mine::TAG
            return Item::Mine.read_from(stream)
        end
        raise "Unexpected discriminant value"
    end

    class HealthPack
        TAG = 0
        attr_accessor :health
        def initialize(health)
            @health = health
        end
        def self.read_from(stream)
            health = stream.read_int()
            HealthPack.new(health)
        end
        def write_to(stream)
            stream.write_int(TAG)
            stream.write_int(@health)
        end
    end
    class Weapon
        TAG = 1
        attr_accessor :weapon_type
        def initialize(weapon_type)
            @weapon_type = weapon_type
        end
        def self.read_from(stream)
            weapon_type = stream.read_int()
            if weapon_type < 0 || weapon_type > 3
                raise "Unexpected discriminant value"
            end
            Weapon.new(weapon_type)
        end
        def write_to(stream)
            stream.write_int(TAG)
            stream.write_int(@weapon_type)
        end
    end
    class Mine
        TAG = 2
        def initialize()
        end
        def self.read_from(stream)
            Mine.new()
        end
        def write_to(stream)
            stream.write_int(TAG)
        end
    end
end

File: clients/java/src/main/java/model/Item.java
package model;

import util.StreamUtil;

public abstract class Item {
    public abstract void writeTo(java.io.OutputStream stream) throws java.io.IOException;
    public static Item readFrom(java.io.InputStream stream) throws java.io.IOException {
        switch (StreamUtil.readInt(stream)) {
            case HealthPack.TAG:
                return HealthPack.readFrom(stream);
            case Weapon.TAG:
                return Weapon.readFrom(stream);
            case Mine.TAG:
                return Mine.readFrom(stream);
            default:
                throw new java.io.IOException("Unexpected discriminant value");
        }
    }

    public static class HealthPack extends Item {
        public static final int TAG = 0;
        private int health;
        public int getHealth() { return health; }
        public void setHealth(int health) { this.health = health; }
        public HealthPack() {}
        public HealthPack(int health) {
            this.health = health;
        }
        public static HealthPack readFrom(java.io.InputStream stream) throws java.io.IOException {
            HealthPack result = new HealthPack();
            result.health = StreamUtil.readInt(stream);
            return result;
        }
        @Override
        public void writeTo(java.io.OutputStream stream) throws java.io.IOException {
            StreamUtil.writeInt(stream, TAG);
            StreamUtil.writeInt(stream, health);
        }
    }

    public static class Weapon extends Item {
        public static final int TAG = 1;
        private model.WeaponType weaponType;
        public model.WeaponType getWeaponType() { return weaponType; }
        public void setWeaponType(model.WeaponType weaponType) { this.weaponType = weaponType; }
        public Weapon() {}
        public Weapon(model.WeaponType weaponType) {
            this.weaponType = weaponType;
        }
        public static Weapon readFrom(java.io.InputStream stream) throws java.io.IOException {
            Weapon result = new Weapon();
            switch (StreamUtil.readInt(stream)) {
            case 0:
                result.weaponType = model.WeaponType.PISTOL;
                break;
            case 1:
                result.weaponType = model.WeaponType.ASSAULT_RIFLE;
                break;
            case 2:
                result.weaponType = model.WeaponType.ROCKET_LAUNCHER;
                break;
            default:
                throw new java.io.IOException("Unexpected discriminant value");
            }
            return result;
        }
        @Override
        public void writeTo(java.io.OutputStream stream) throws java.io.IOException {
            StreamUtil.writeInt(stream, TAG);
            StreamUtil.writeInt(stream, weaponType.discriminant);
        }
    }

    public static class Mine extends Item {
        public static final int TAG = 2;
        public Mine() {}
        public static Mine readFrom(java.io.InputStream stream) throws java.io.IOException {
            Mine result = new Mine();
            return result;
        }
        @Override
        public void writeTo(java.io.OutputStream stream) throws java.io.IOException {
            StreamUtil.writeInt(stream, TAG);
        }
    }
}
